### ISSUE TITLE
Simple example even simpler

### ISSUE THREAD

**Issue by jan-janssen** [pyiron/pyiron_workflow#35]
**Summary**

Currently, the `pyiron_workflow` code is still more complex than other frameworks.

**Detailed Description**

Example from the `README.MD`: 
```
wf = Workflow("my_workflow")
wf.a1 = x_plus_y()
wf.a2 = x_plus_y()
wf.b = x_plus_y(x=wf.a1.outputs.sum, y=wf.a2.outputs.sum)

out = wf(a1__x=0, a1__y=1, a2__x=2, a2__y=3)
out.b__sum
>>> 6
```

I would like to write: 
```
out = x_plus_y(
    x=x_plus_y(x=0, y=1),
    y=x_plus_y(x=2, y=3),
)
out.sum
```

**Further Information, Files, and Links**

As far as I understand this is currently related to enforcing labels for the individual nodes of the workflow, so the suggestion would be to generate labels automatically. I am aware that labels can be helpful for debugging the workflow, but I would like to have a certain level of compatibility to existing python functions.

**Comment by liamhuber** [comment]
The demo is intentionally not the absolute simplest thing possible to demonstrate there is more flexibility and power than what you are showing. You can, of course, use the nodes without putting them in a workflow first. 

There are a couple fundamental problems with the syntax you're requesting:
1. Execution is intentionally delayed, so `out = x_plus_y(...); out.sum` is missing an invocation to run things. Joerg and I have discussed this at length and settled that delayed execution is definitely desirable. It's possible there's a way to implement your desired syntax in a delayed way, but I don't see it -- we want to be able to both instantiate nodes and inspect their output without forcing an execution.
2. In principle, we allow nodes to have multiple return values, so argument assignments like `x=x_plus_y(x=0, y=1)` are not (necessarily*) sufficient
3. You _can_ instantiate new nodes inside a function call, but IMO it's bad practice since then you do not have easy reference to those node objects. So this you can already do, I just don't advise it.

The "necessarily" above is that we have a separate `SingleValue(Function)` class that lets you just stick the entire node in as an output channel, since it only has one of these and thus is completely unambiguous. The snippet you posted is incomplete, but making this change is as simple as replacing `@Workflow.wrap_as.function_node` with `@Workflow.wrap_as.single_value_node`

Here is an example that is headed in the same direction as what you show and runs on the current code:
```
@Workflow.wrap_as.single_value_node("sum")
def x_plus_y(x: int = 0, y: int = 0) -> int:
    return x + y

n1 = x_plus_y(0, 1)
n2 = x_plus_y(2, 3)
n3 = x_plus_y(n1, n2)
n1 > n2 > n3 
n1() 
n3.outputs.sum.value
>>> 6
```

In workflows we do topological analysis on data connections to automatically define execution flows. Using the exact same tools for nodes that may or may not be inside a workflow by invoking a `pull()` method is what I've been working towards this week and part of last week, and is discussed more verbosely in #12 and #10, which are being implemented in #23 (e.g. [this commit](https://github.com/pyiron/pyiron_workflow/pull/23/commits/ee4fd8adf9f46174aaa3f78f62603dab806189f1) which allows input panels to fetch the most recent upstream data). 

Because we don't yet expose tools to exploit data topology outside a workflow context, we're still stuck manually defining and triggering execution with the lines `n1 > n2 > n3; n1()`, which also necessitates that we instantiate all the nodes and have variable for them ahead of time. I agree this is ugly. A successful implementation of `pull()` will allow us to write the following:

```
@Workflow.wrap_as.single_value_node("sum")
def x_plus_y(x: int = 0, y: int = 0) -> int:
    return x + y

node = x_plus_y(x=x_plus_y(0, 1), y=x_plus_y(2, 3), 1))
node.pull()
>>> 6
# Pull should return the value directly,
# just like running a function node currently does
# Of course it's still available later at 
# node.outputs.sum.value
```

Although I still don't really _advise_ instantiating nodes as arguments for other nodes, because if you want to reference them in the future you'd need to track it down with something some ugly like `nodex = node.inputs.x.connections[0].node` -- usually wiser to just instantiate it into it's own variable and then write `node = x_plus_y(nodex, nodey)` (or whatever you named the variables). If you _really_ don't want to have any node variables kicking around, you could even do the above as a one-liner: `x_plus_y(x=x_plus_y(0, 1), y=x_plus_y(2, 3), 1)).pull()`

That's all I have to say about easy syntax and workflow-free running, as a final aside on this comment:

> As far as I understand this is currently related to enforcing labels for the individual nodes of the workflow, so the suggestion would be to generate labels automatically. 

I just want to point out that we already do generate labels automatically, not just for function nodes but also for workflows. If you just add a node to a workflow a name will be generated for it automatically based on the node class. If you turn off strict naming you can even add things repeatedly and it will just quietly iterate on the class name:

```
@Workflow.wrap_as.single_value_node("sum")
def x_plus_y(x: int = 0, y: int = 0) -> int:
    return x + y

wf = Workflow("my_wf", strict_naming=False)
wf.add(x_plus_y(0, 1))
wf.add(x_plus_y(2, 3))
wf.add(x_plus_y(wf.x_plus_y, wf.x_plus_y0))
wf.nodes.keys()
>>> dict_keys(['x_plus_y', 'x_plus_y0', 'x_plus_y1'])
```

**Comment by jan-janssen** [comment]
Hi @liamhuber, looking through the examples I guess the `pull` mode is exactly what I am looking for.
